FFmpegInterop/UncompressedVideoSampleProvider.cpp

```diff
@@ -18,140 +18,111 @@
-	for (int i = 0; i < 4; i++)
-	{
-		m_rgVideoBufferLineSize[i] = 0;
-		m_rgVideoBufferData[i] = nullptr;
-	}
-}
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Media::Core;
+using namespace winrt::Windows::Media::MediaProperties;
+using namespace winrt::Windows::Storage::Streams;
+using namespace std;
 
-HRESULT UncompressedVideoSampleProvider::AllocateResources()
+namespace winrt::FFmpegInterop::implementation
 {
-	HRESULT hr = S_OK;
-	hr = UncompressedSampleProvider::AllocateResources();
-	if (SUCCEEDED(hr))
+	UncompressedVideoSampleProvider::UncompressedVideoSampleProvider(_In_ AVStream* stream, _In_ Reader& reader) :
+		UncompressedSampleProvider(stream, reader)
 	{
-		// Setup software scaler to convert any decoder pixel format (e.g. YUV420P) to NV12 that is supported in Windows & Windows Phone MediaElement
-		m_pSwsCtx = sws_getContext(
-			m_pAvCodecCtx->width,
-			m_pAvCodecCtx->height,
-			m_pAvCodecCtx->pix_fmt,
-			m_pAvCodecCtx->width,
-			m_pAvCodecCtx->height,
-			AV_PIX_FMT_NV12,
-			SWS_BICUBIC,
-			NULL,
-			NULL,
-			NULL);
-
-		if (m_pSwsCtx == nullptr)
+		if (m_codecContext->pix_fmt != AV_PIX_FMT_NV12)
 		{
-			hr = E_OUTOFMEMORY;
+			// Setup software scaler to convert the pixel format to NV12
+			m_swsContext.reset(sws_getContext(
+				m_codecContext->width,
+				m_codecContext->height,
+				m_codecContext->pix_fmt,
+				m_codecContext->width,
+				m_codecContext->height,
+				AV_PIX_FMT_NV12,
+				SWS_BICUBIC,
+				nullptr,
+				nullptr,
+				nullptr));
+			THROW_IF_NULL_ALLOC(m_swsContext);
+
+			THROW_HR_IF_FFMPEG_FAILED(av_image_fill_linesizes(m_lineSizes, AV_PIX_FMT_NV12, m_codecContext->width));
+
+			// Create a buffer pool
+			const int requiredBufferSize = av_image_get_buffer_size(AV_PIX_FMT_NV12, m_codecContext->width, m_codecContext->height, 1);
+			THROW_HR_IF_FFMPEG_FAILED(requiredBufferSize);
+
+			m_bufferPool.reset(av_buffer_pool_init(requiredBufferSize, nullptr));
+			THROW_IF_NULL_ALLOC(m_bufferPool);
 		}
 	}
 
-	if (SUCCEEDED(hr))
+	void UncompressedVideoSampleProvider::SetEncodingProperties(_Inout_ const IMediaEncodingProperties& encProp, _In_ bool setFormatUserData)
 	{
-		m_pAvFrame = av_frame_alloc();
-		if (m_pAvFrame == nullptr)
+		SampleProvider::SetEncodingProperties(encProp, setFormatUserData);
+
+		VideoEncodingProperties videoEncProp{ encProp.as<VideoEncodingProperties>() };
+
+		if (m_codecContext->framerate.num != 0 && m_codecContext->framerate.den != 0)
 		{
-			hr = E_OUTOFMEMORY;
+			MediaRatio frameRate{ videoEncProp.FrameRate() };
+			frameRate.Numerator(m_codecContext->framerate.num);
+			frameRate.Denominator(m_codecContext->framerate.den);
 		}
+
+		MediaPropertySet videoProp{ videoEncProp.Properties() };
+		videoProp.Insert(MF_MT_INTERLACE_MODE, PropertyValue::CreateUInt32(MFVideoInterlace_MixedInterlaceOrProgressive));
 	}
 
-	if (SUCCEEDED(hr))
+	tuple<IBuffer, int64_t, int64_t, map<GUID, IInspectable>> UncompressedVideoSampleProvider::GetSampleData()
```

Comment:
>GetSampleData [](start = 92, length = 13)

Looks like we don't handle format changes (i.e resolution changes). For cases where there frame width / height changes and we're not performing a sws_scale operation (as the pixel format is already NV12), we would send sample data which doesn't conform to the media type exposed on the stream descriptor. And if we are performing a sws_scale operation, we would end up scaling up/down to the old resolution. Neither of these cases are ideal and AFAICT we'd need changes on the OS side in MSS to properly support format changes. For now, we may want to at least prevent the case where we send samples that don't conform to the media type by dynamically initializing the m_swsContext if the frame width/height doesn't match the original width/height (this could probably also be done in a follow up PR since we didn't handle this correctly prior to the refactor either).

Reply:
MSS already supports format changes - on each sample request we check to see if the stream descriptor's encoding properties have been modified. I'll make the necessary changes to handle this case.